### PR TITLE
Delay evaluation of resource.Default to first call

### PR DESCRIPTION
**Pull request by bmon**
Currently `defaultResource` (which includes data calculated by parsing env var `OTEL_RESOURCE_ATTRIBUTES`) is evaluated as an import side effect. Some applications might want to update that variable with go code (eg to load extra env vars from a configuration file, for example using https://github.com/joho/godotenv). By only evaluating it the first time `resource.Default()` is called, we can allow those applications an opportunity to set `OTEL_RESOURCE_ATTRIBUTES` first, while still only calculating the value once.

### CODE REVIEW

File: sdk/resource/resource.go [hunk 1]
```diff
@@ -211,6 +196,18 @@ func Empty() *Resource {
 // Default returns an instance of Resource with a default
 // "service.name" and OpenTelemetrySDK attributes.
 func Default() *Resource {
+	if defaultResource == nil {
```
Comment (Aneurysm9) [inline sdk/resource/resource.go:199]: It looks like it is possible for the return value of `Detect` to be nil.  At the end of this block please set `defaultResource = emptyResource` if that happens.  It should never happen with the set of detectors provided here but things change and that could cause this detection to run more than once per process.

Comment (evantorrie) [reply sdk/resource/resource.go:199]: May want to also consider (overkill perhaps?) using a `sync.Once` to avoid races...

**Review by jmacd — APPROVED** [review]

**Review by MrAlias — COMMENTED** [review]
The implementation looks good to me. Can you add an entry to the changelog describing the changed behavior to the end user?

**Review by evantorrie — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Comment by bmon** [comment]
Thanks very much, all. Cheers! 👍

**Pull request merged**